R/backend-file.R

```diff
@@ -0,0 +1,507 @@
+
+  assert_that(file.exists(private$keyring_file(keyring)))
+
+  private$key_unset(keyring)
+
+  invisible(self)
+}
+
+b_file_keyring_unlock <- function(self, private, keyring, password) {
+
+  private$key_set(password, keyring)
+
+  assert_that(file.exists(private$keyring_file(keyring)))
+
+  if (self$keyring_is_locked(keyring)) {
+    private$key_unset(keyring)
+    b_file_error("failed to unlock keyring")
+  }
+
+  invisible(self)
+}
+
+b_file_keyring_is_locked <- function(self, private, keyring) {
+
+  keyring <- keyring %||% private$keyring
+
+  if (!file.exists(keyring) || !private$key_is_set(keyring)) {
+    TRUE
+  } else {
+    tryCatch(
+      {
+        b_file_secret_decrypt(
+          private$check_get(keyring),
+          private$nonce_get(keyring),
+          private$key_get(keyring)
+        )
+        FALSE
+      },
+      error = function(e) {
+        if(conditionMessage(e) == "Failed to decrypt")
+          TRUE
+        else
+          signalCondition(e)
+      }
+    )
+  }
+}
+
+b_file_keyring_default <- function(self, private) {
+  private$keyring
+}
+
+b_file_keyring_set_default <- function(self, private, keyring) {
+  private$keyring <- keyring
+  invisible(self)
+}
+
+## --------------------------------------------------------------------
+## Private
+
+b_file_keyring_create_direct <- function(self, private, keyring, password,
+  nonce, items) {
+
+  file_name <- keyring %||% private$keyring
+
+  if (file.exists(file_name))
+    confirmation(paste("are you sure you want to overwrite", file_name))
+
+  private$keyring_write_file(
+    file_name,
+    nonce %||% sodium::random(24L),
+    items %||% list(),
+    password %||% get_pass()
+  )
+
+  invisible(self)
+}
+
+b_file_keyring_file <- function(self, private, keyring, ...) {
+
+  file_name <- keyring %||% private$keyring
+
+  assert_that(is_string(file_name))
+
+  if (!file.exists(file_name))
+    private$keyring_create_direct(file_name, ...)
+
+  normalizePath(file_name, mustWork = TRUE)
+}
+
+b_file_read_keyring_file <- function(self, private, keyring) {
+
+  yml <- yaml::yaml.load_file(keyring %||% private$keyring)
+
+  assert_that(is.list(yml),
+              assertthat::has_name(yml, "keyring_info"),
+              is.list(yml[["keyring_info"]]),
+              assertthat::has_name(yml[["keyring_info"]], "nonce"),
+              assertthat::has_name(yml[["keyring_info"]], "integrity_check"),
+              assertthat::has_name(yml, "items"),
+              is.list(yml[["items"]]))
+
+  list(
+    nonce = sodium::hex2bin(yml[["keyring_info"]][["nonce"]]),
+    items = yml[["items"]],
+    check = yml[["keyring_info"]][["integrity_check"]]
+  )
+}
+
+b_file_write_keyring_file <- function(self, private, keyring, nonce, items,
+  key) {
+
+  nonce <- nonce %||% private$nonce_get(keyring)
+
+  yaml::write_yaml(
+    list(
+      keyring_info = list(
+        keyring_version = as.character(
+          utils::packageVersion(methods::getPackageName())
```

Comment:
You need to import the methods package for this.  Or, preferably, just use `.packageName`.

Reply:
done (didn't know about this, thanks!).